dashboard/src/modules/components/DatePickerComponent/index.jsx

```diff
@@ -0,0 +1,83 @@
+import React, { useState } from "react";
+import {
+  InputGroup,
+  InputGroupText,
+  DatePicker,
+  isValidDate,
+  Button,
+} from "@patternfly/react-core";
+import "./index.css";
+import { formatDate } from "../../../utils/dateFormatter";
+import { filterData } from "../../../utils/filterDataset";
+import { constructUTCDate } from "../../../utils/constructDate";
+import bumpToDate from "utils/bumpDate";
+import addOneDay from "utils/addDay";
+
+function DatePickerWidget({
+  dataArray,
+  setPublicData,
+  controllerName,
+  setDateRange,
+}) {
+  const [fromDate, setFromDate] = useState({});
+  const [toDate, setToDate] = useState(
+    constructUTCDate(formatDate(new Date(bumpToDate(new Date(), 1))))
+  );
+  const [strDate, setStrDate] = useState(
+    new Date().toLocaleString("fr-CA").split(",")[0]
+  );
+  const toValidator = (date) =>
+    date >= fromDate
+      ? ""
+      : "To date must be greater than or equal to from date";
+
+  const onFromChange = (_str, date) => {
+    const selectedDate = constructUTCDate(_str);
+    setFromDate(isValidDate(date) ? selectedDate : {});
+    if (isValidDate(date)) {
+      if (date > new Date(strDate)) {
+        setToDate(
+          constructUTCDate(formatDate(new Date(bumpToDate(new Date(_str), 2))))
```

Comment:
Two days?

Reply:
My Idea (which I tried to explain in design sync,but I was sure I wouldn’t be able to explain) was we want the toDate to be pointing to the **end** of midnight(rather than start of midnight in case of from date),so I need to bump the selected todate values.Now,if I am selecting the toDate from date picker itself,then bumping up by one day is sufficient,but in this case,_str is the selected fromDate,so I need to bump the value by 2

(Makes Sense?No?)

Reply:
I'm guessing that what is happening here is that @shubham-html-css-js wants the displayed to-date to be different from the displayed from-date, but, since we're going to truncate each of the date-time values back to midnight-at-the-beginning-of-the-day, he needs to move the date forward two days so that it will still be different when it's truncated back to the beginning of the day.

My thinking on this is that if the from-date and the to-date are the same (on the user's screen) that's fine, so long as the internals conspire to make that "range" include everything between midnight-at-the-beginning-of-the-day and midnight-at-the-end-of-the-day; so, therefore, the to-date doesn't have to be bumped by more than one day; and, whether we display that as the same day as the from-date or as the next day is a detail which should be handled separately from the actual value of the to-date.

Reply:
@shubham-html-css-js, I think you and I were composing at the same time.  🙂

Does my reply change your thinking?  (Or, should I be trying to change mine?  😀)

Reply:
I'm not sure what "the end of midnight" means. Sounds like a movie title.

I think that, if "now" is 2022-05-05T19:15 and we're defaulting to "the current day" as a search range, we want "from" to be 2022-05-06T00:00 and "to" to be 2022-05-06T00:00. Neither of those are 2 days, so I clearly don't understand the context of your explanation.

Reply:
I will try to explain my thought process in a detailed way

I am using `bumpToDate` function at three places in my code,out of which I will quote the use of it at 2 places,which are:-

1)When a user manually selects date from toDate Picker
2)When a user selects a fromDate,greater than toDate

Now,we want the fromDate to be pointing to midnight(start of day,which we fortunately get by default) and toDate to be pointing to midnight(end of day),so that all dates are encapsulated between those ranges.We know that by default what we get from patternfly is midnight at the start of day,which will cause problem when it is normalized(truncated back),and we might miss out on some dates.That's where the concept of bumping up the dates comes into Play.

When a user selects some day from patternfly or if we are defaulting to the current day for our search range,we will bump the date by one day(which is what I am doing if you see the useState() initialization and toDate on change method).But when a user selects a fromDate which is greater than the toDate,I in the fromDate onChange Listener has only access to the selected fromDate and not the toDate.So ,if I bump the fromDate by just 1 day,it will be the end-of-day midnight  for the fromDate,which is equivalent to start of day midnight for the toDate,so that is why I am bumping the from Date to 2 days,so that we are pointing to end-of-day midnight toDate in this scenario also and we don't loose any dates.

(Does this makes sense?)

Reply:
> if I bump the fromDate by just 1 day,it will be the end-of-day midnight for the fromDate,which is equivalent to start of day midnight for the toDate,so that is why I am bumping the from Date to 2 days,so that we are pointing to end-of-day midnight toDate in this scenario also and we don't loose any dates.
> 
> (Does this makes sense?)

We need to be really careful here, otherwise we won't be able to avoid confusion -- we need to separate this into two issues:  the first is what the internal date-range is for selecting datasets; the second is what values we receive from and display to the user in the `DatePicker`.

For the internal date range, for the case where the user selects a from-date which is larger than the to-date, I think we want the to-date to be reset to midnight (UTC) at the beginning of the day after the from-date.  This will produce a range encompassing 24 hours, and any dataset timestamps (which now have 1 second resolution) which fall in that interval will be selected.

However, this should be decoupled from what we then display to the user.  That is, we need to decide how we represent the to-date for the user, whether we display it as the same date as the from-date (with the understanding that the range encompasses exactly the whole day) or whether we display it as the next day (with the understanding that the actual bounds are their respective beginning-of-day-midnights).

In the context of the code snippet above, I think we are talking about the internal date range.  Therefore, we need to move only one day forward from the from-date to set the to-date.

(Once we understand how to represent the internal range, then we can move on to discuss what the user should see and how to make that happen...but lets hold off on that.)

Reply:
Yes,I agree on the above point.It feels like we are on a same page now.

What I am trying to do is rather then encompassing 24 hrs,I am trying to strive for encompassing 48 hrs,since from the beginning ,my intention was if the user selects a fromDate greater then toDate,then two date must shift one day ahead ,and if that is the case ,then encapsulating 48 hrs makes sense.On the other hand,if I change the code to bump it by one day,then it will be a 24 hr encapsulation in which the fromDate and toDate would be the same.I want to go for 48 hrs encapsulation,since I feel that would be more logical(or maybe what I personally would like as a user?).

But I am glad that I was able to explain the point.Now it feels like we are close to a merge.(Though number of unresolved conversations have increased ,still I guess we now have addressed the elephant in the room and worked on that)

Reply:
> What I am trying to do is rather then encompassing 24 hrs,I am trying to strive for encompassing 48 hrs,since from the beginning ,my intention was if the user selects a fromDate greater then toDate,then two date must shift one day ahead ,and if that is the case ,then encapsulating 48 hrs makes sense.On the other hand,if I change the code to bump it by one day,then it will be a 24 hr encapsulation in which the fromDate and toDate would be the same.I want to go for 48 hrs encapsulation,since I feel that would be more logical(or maybe what I personally would like as a user?).

@shubham-html-css-js, I think you are combining two issues that we should keep separated.  Let me take another tack.

What if, I as a user, want to select datasets from a single day, say, 2022-05-04 (UTC) -- what dates should I select in the `DatePicker`?

If I select a from-date of 2022-05-04, then I won't see any datasets with dates from before that.  So far, so good.  But, if a select a to-date of 2022-05-05, will I see datasets from the 5th?  (I don't want to see those.)  If I select a to-date of 2022-05-04, will I see no datasets (because the "from" and "to" values are "the same")?

The first point I want to make here is that the user experience is a separate problem from the internal code's selecting the list of datasets.  The two are interconnected, to be sure, but the problems should be managed separately.

If you think that the user experience should be that the to-date is always strictly greater than the from-date, then that has certain implications, but it doesn't mean that the internal representation of the to-date must be more than 24 hours larger than the from-date.  The code will take what the `DatePicker` conveys from the user and it will transform it as necessary.  We need to decide what the user experience should be and what the significance of the selections are.  That will drive the transformations that the code will perform, but it doesn't dictate how the internals work.

> Though number of unresolved conversations have increased, still I guess we now have addressed the elephant in the room and worked on that

The number of conversations is up from 11 to 12, but I think at least 3 of them deal with the same issue in different parts of the code, and I think there is at least one other pair which overlap or are duplicates, so it's really more like 8 issues.

As for the "elephant in the room", I'm not sure we've addressed it as much as all realized that it's there.  🐘

Reply:
@dbutenhof, as far as I'm concerned, this conversation can be resolved.